src/backends/mod.rs

```diff
@@ -1,38 +1,25 @@
 use super::types::{Message, Partition, Position, Topic};
 use std::collections::{HashMap, HashSet};
+use thiserror::Error;
 
 pub mod kafka;
 pub mod local;
 pub mod storages;
 
-#[derive(Debug, Clone)]
-pub struct ConsumerClosed;
-
-#[derive(Debug, Clone)]
-pub struct EndOfPartition;
-
-#[derive(Debug, Clone)]
-pub enum PollError {
-    ConsumerClosed,
+#[derive(Error, Debug)]
+pub enum ConsumerError {
+    #[error("End of partition reached")]
     EndOfPartition,
-}
-
-#[derive(Debug, Clone)]
-pub struct UnassignedPartition;
 
-#[derive(Debug, Clone)]
-pub enum PauseError {
+    #[error("The consumer is closed")]
     ConsumerClosed,
-    UnassignedPartition,
-}
 
-#[derive(Debug, Clone)]
-pub struct ConsumerError;
+    #[error("Partition not assigned to consumer")]
+    UnassignedPartition,
 
-#[derive(Debug, Clone)]
-pub enum ConsumeError {
-    ConsumerClosed,
-    ConsumerError,
+    // Pretty much any useful error.
+    #[error(transparent)]
+    Other(#[from] Box<dyn std::error::Error>),
```

Comment:
Yes, 
The abstract consuemr does not know the specific concrete errors of the rdkafka implementation. For as long as this is a trait we cannot change this. If we gave up on having an abstract trait and made the interface Kafka specific as well we could change it. 

Though that is something to think carefully about. Specifically now that there is a push to support RedPanda

Reply:
> I think the right approach is probably to map each possible type of error (or error code) we can get back from rdkafka to one of the actual errors on this list (and ideally never use "other")

This would be the list: https://github.com/fede1024/rust-rdkafka/blob/40598f5a39d74bbf2298c1dbeeb0c1ae2fb0919e/src/error.rs#L135-L182
which then map these 
https://github.com/fede1024/rust-rdkafka/blob/5d23e82a675d9df1bf343aedcaa35be864787dab/rdkafka-sys/src/types.rs#L110-L113

I think we should seriously think through whether we should keep the arroyo consumer kafka agnostic (which is not really already). Will look into the Redpanda interface (which is supposed to be Kafka compatible). 
If we map all Kafka specific errors we would be making the consumer api kafka specific.

Reply:
Well, this is the redpanda doc -> use the same kafka libraries 
https://docs.redpanda.com/docs/reference/faq/#what-are-the-recommended-clients-to-use-with-redpanda

Reply:
> I think we should seriously think through whether we should keep the arroyo consumer kafka agnostic (which is not really already). Will look into the Redpanda interface (which is supposed to be Kafka compatible). If we map all Kafka specific errors we would be making the consumer api kafka specific.

I'm not necessarily suggesting making the errors Kafka specific at all, or that the mapping needs to be 1:1. I would be fine with mapping most of them to a fairly generic "ConsumerError" or something. But at least for the ones that we will have defined elsewhere on this list (like OffsetOutOfRange, TopicDoesNotExist, etc) I think we need to map it to the correct error, not Other.